src/test/Profiling.cpp

```diff
@@ -0,0 +1,230 @@
+
+using namespace std;
+using namespace std::chrono_literals;
+using namespace crepe;
+using namespace testing;
+
+
+/*
+List of test cases with component settings/details
+1. Minimal test creates gameobject without additonal components
+2. Minimal 'Complex' gameobject. Has dynamic body without bounce and no collision handler
+3. Minimal 'Complex' gameobject. Same as test 2 but with particle emitter
+*/
+
+class TestScript : public Script {
+	bool oncollision(const CollisionEvent& test) {
+		Log::logf("Box {} script on_collision()", test.info.first_collider.game_object_id);
+		return true;
+	}
+	void init() {
+		subscribe<CollisionEvent>([this](const CollisionEvent& ev) -> bool {
+			return this->oncollision(ev);
+		});
+	}
+	void update() {
+		// Retrieve component from the same GameObject this script is on
+	}
+};
+
+class Profiling : public Test {
+public:
+	// Config for test
+	// Minimum amount to let test pass
+	const int min_gameobject_count = 100;
+	// Maximum amount to stop test
+	const int max_gameobject_count = 200;
+	// Amount of times a test runs to calculate average
+	const int average = 5;
+	// Maximum duration to stop test
+	const std::chrono::microseconds duration = 16000us;
+	
+
+	ComponentManager mgr;
+	// Add system used for profling tests 
+	CollisionSystem collision_sys{mgr};
+	PhysicsSystem physics_sys{mgr};
+	ParticleSystem particle_sys{mgr};
+	RenderSystem render_sys{mgr};
+	ScriptSystem script_sys{mgr};
+
+	// Test data
+	std::map<std::string, std::chrono::microseconds> timings;
+	int game_object_count = 0;
+	std::chrono::microseconds total_time = 0us;
+
+	
+	void SetUp() override {
+		
+		GameObject do_not_use = mgr.new_object("DO_NOT_USE","",{0,0});
+		do_not_use.add_component<Camera>(Color::WHITE);
+		// initialize systems here:
+		//calls init
+		script_sys.update();
+		//creates window
+		render_sys.update();
+	}
+
+	// Helper function to time an update call and store its duration
+	template <typename Func>
+	std::chrono::microseconds time_function(const std::string& name, Func&& func) {
+		auto start = std::chrono::steady_clock::now();
+		func();
+		auto end = std::chrono::steady_clock::now();
+		std::chrono::microseconds duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
+		timings[name] += duration;
+		return duration;
+	}
+
+	// Run and profile all systems, return the total time in milliseconds
+	std::chrono::microseconds run_all_systems() {
+		std::chrono::microseconds total_microseconds = 0us;
+		total_microseconds += time_function("PhysicsSystem", [&]() { physics_sys.update(); });
+		total_microseconds += time_function("CollisionSystem", [&]() { collision_sys.update(); });
+		total_microseconds += time_function("ParticleSystem", [&]() { particle_sys.update(); });
+		total_microseconds += time_function("RenderSystem", [&]() { render_sys.update(); });
+		return total_microseconds;
+	}
+
+	// Print timings of all functions
+	void log_timings() const {
+		std::stringstream ss;
```

Comment:
Liever gewoon een `std::string` en `std::format`, deze code is nu heel slecht leesbaar